packages/host/app/services/render-service.ts

```diff
@@ -22,27 +50,108 @@ export default class RenderService extends Service {
   renderError: Error | undefined;
   owner: Owner = getOwner(this)!;
 
-  async renderCard(
-    url: URL,
-    staticResponses: Map<string, string>
-  ): Promise<string> {
-    this.renderError = undefined;
-    this.loaderService.setStaticResponses(staticResponses);
-    let card = await this.cardService.loadModel(url);
-    if (!card) {
-      throw new Error(`card ${url.href} not found`);
-    }
-    let component = card.constructor.getComponent(card, 'isolated', {
+  async renderCard(params: RenderCardParams): Promise<string> {
+    let {
+      card,
+      visit,
+      format = 'embedded',
+      identityContext,
+      realmPath,
+    } = params;
+    let component = card.constructor.getComponent(card, format, {
       disableShadowDOM: true,
     });
 
-    let serializer = new Serializer(voidMap);
     let element = getIsolatedRenderElement(this.document);
+    // this loop looks really similar to the API's recompute's
+    // _loadModel(). The primary difference is that we are letting the render
+    // drive the field load as opposed to the card schema
+    let notReady: NotReady | undefined;
+    let notLoaded: NotLoaded | undefined;
+    do {
+      notReady = undefined;
+      notLoaded = undefined;
+      try {
+        render(component, element, this.owner);
+      } catch (err: any) {
+        notReady = err.additionalErrors?.find((e: any) =>
+          isNotReadyError(e)
+        ) as NotReady | undefined;
+        notLoaded = err.additionalErrors?.find((e: any) =>
+          isNotLoadedError(e)
+        ) as NotLoaded | undefined;
+        if (isCardError(err) && (notReady || notLoaded)) {
+          card = notReady?.model ?? card; // a nested card may be the card that is not ready
+          let fieldName = notReady?.fieldName ?? notLoaded?.fieldName;
+          if (!fieldName) {
+            throw new Error(`bug: should never get here`);
+          }
+          await this.resolveField({
+            card,
+            fieldName,
+            identityContext,
+            visit,
+            realmPath,
+          });
+        } else {
+          throw err;
+        }
+      }
 
-    render(component, element, this.owner);
+      // TODO make a test that shows we throw for fields that can't be loaded (broken links)
+    } while (notLoaded || notReady);
```

Comment:
Add additional logic to assert that we don't encounter the same NotReady/NotLoaded on this pass as we did in the previous pass--when there are bugs, this causes a never ending loop because we get trapped by the same NotReady/NotLoaded on each pass.